src/__tests__/span_processor.test.ts

```diff
@@ -0,0 +1,157 @@
+  let tracerProvider: BasicTracerProvider
+  let client: Client
+
+  function mockClient() {
+    const originalCreateOpenTelemetrySpan =
+      client.extension.createOpenTelemetrySpan
+
+    jest
+      .spyOn(client.extension, "createOpenTelemetrySpan")
+      .mockImplementation((...args) => {
+        const span = originalCreateOpenTelemetrySpan(...args)
+        mockSpan(span)
+        return span
+      })
+  }
+
+  function mockSpan(span: Span) {
+    const originalClose = span.close.bind(span)
+
+    jest.spyOn(span, "close").mockImplementation(() => {
+      const storedToObject = {
+        ...span.toObject(),
+        closed: true
+      }
+
+      jest.spyOn(span, "toObject").mockImplementation(() => storedToObject)
+
+      return originalClose()
+    })
+  }
+
+  function createdSpans() {
+    return (
+      client.extension.createOpenTelemetrySpan as jest.Mock
+    ).mock.results.map(result => result.value.toObject())
+  }
+
+  function createError(message: string): any {
+    try {
+      throw new Error(message)
+    } catch (e) {
+      return e
+    }
+  }
+
+  beforeEach(() => {
+    client = new Client({
+      name: "TEST APP",
+      pushApiKey: "PUSH_API_KEY",
+      active: true,
+      enableMinutelyProbes: false
+    })
+
+    mockClient()
+
+    const spanProcessor = new SpanProcessor(client)
+
+    tracerProvider = new BasicTracerProvider()
+    tracerProvider.addSpanProcessor(spanProcessor)
+
+    tracer = tracerProvider.getTracer("unknown-instrumentation")
+  })
+
+  afterEach(() => {
+    tracerProvider.shutdown()
+    client.stop()
+  })
+
+  it("processes unknown OpenTelemetry spans", () => {
+    tracer.startSpan("unknownSpan").setAttribute("foo", "bar").end()
+
+    const spans = createdSpans()
+    expect(spans.length).toEqual(1)
+    expect(spans[0]).toMatchObject({
+      name: "unknownSpan",
+      closed: true,
+      attributes: expect.objectContaining({
+        foo: "bar",
+        "appsignal:body": expect.stringContaining("unknown-instrumentation")
+      }),
+      parent_span_id: "",
+      error: null
+    })
+  })
+
+  it("processes the first exception event in an OpenTelemetry span as the error", () => {
+    const span = tracer.startSpan("errorSpan")
+    span.recordException(createError("first"))
+    // the following error will be ignored
+    span.recordException(createError("last"))
+    span.end()
+
+    const spans = createdSpans()
+    expect(spans.length).toEqual(1)
+    expect(spans[0]).toMatchObject({
+      name: "errorSpan",
+      closed: true,
+      parent_span_id: "",
+      error: {
+        name: "Error",
+        message: "first",
+        backtrace: expect.any(String)
+      }
+    })
+
+    const backtrace = JSON.parse(spans[0].error.backtrace)
+    expect(backtrace).toBeInstanceOf(Array)
+
+    expect(backtrace[0]).toEqual("Error: first")
+    expect(backtrace[1]).toMatch(/^\s+at createError \(/)
+  })
+
+  it("ignores exception events in child spans", async () => {
```

Comment:
Didn't we say the last error would be stored on the span instead of the first? Or does OpenTelemetry work different in this area of error handling?

Reply:
I thought so too! I've written the test for what the currently existing behavior is.

Looking at it now, the test is kind of a lie: all errors _are_ being saved to the span, but the test can only assert the last one due to [a limitation of the `SpanInProgress.to_json()` method on the agent](https://github.com/appsignal/appsignal-agent/blob/d573c9b59bfadb546354cd995827e87e74d190ef/extension/src/tracing/span.rs#L440-L447) used for the assertions, which will only output the first error.

I'll look for a way to address this.